src/query/sql/src/planner/optimizer/optimizer.rs

```diff
@@ -480,32 +480,20 @@ async fn optimize_merge_into(mut opt_ctx: OptimizerContext, s_expr: SExpr) -> Re
         })
     }
 
-    plan.change_join_order = join_order_changed;
-    if opt_ctx.enable_distributed_optimization {
-        let merge_source_optimizer = MergeSourceOptimizer::create();
-        // Inner join shouldn't add `RowNumber` node.
-        let mut enable_right_broadcast = false;
-
-        if matches!(join_op.join_type, JoinType::RightAnti | JoinType::Right)
-            && merge_source_optimizer
-                .merge_source_matcher
-                .matches(&join_s_expr)
+    let distributed = !join_s_expr.has_merge_exchange();
+    if opt_ctx.enable_distributed_optimization && distributed {
+        let merge_optimizer = MergeOptimizer::create();
+        // If inner join && !join_order_changed, source will broadcast,
+        // others use shuffle.
+        if matches!(
+            join_op.join_type,
+            JoinType::Left | JoinType::LeftAnti | JoinType::Inner if join_order_changed
```

Comment:
If there is a broadcast join and the merge type is `MatchedOnly`, maybe we can remove the shuffle processor(after join, shuffle by row_id) and keep the update data in the local node, because the target data is divided by partition.